oauth2_http/java/com/google/auth/oauth2/CertificateIdentityPoolSubjectTokenSupplier.java

```diff
@@ -97,33 +103,153 @@ private static String encodeCert(X509Certificate certificate)
+    // certificate loading and parsing. These exceptions are wrapped in a new IOException,
+    // as declared by this method's signature.
+    catch (NoSuchFileException e) {
+      // Handles the case where the leaf certificate file itself cannot be found.
+      throw new IOException(String.format("Leaf certificate file not found: %s", leafCertPath), e);
     } catch (CertificateException e) {
-      // Catch CertificateException to provide a more specific error message including
-      // the path of the file that failed to parse, and re-throw as IOException
-      // as expected by the getSubjectToken method signature for I/O related issues.
+      // Handles errors during the parsing of certificate data, which could stem from
+      // issues in either the leaf certificate or the trust chain. The message includes
+      // paths to both for comprehensive error reporting.
       throw new IOException(
-          "Failed to parse certificate(s) from: " + credentialSource.getCredentialLocation(), e);
+          "Failed to read certificate file(s). Leaf path: "
+              + leafCertPath
+              + (trustChainPath != null ? "\nTrust chain path: " + trustChainPath : ""),
+          e);
     }
   }
+
+  /**
+   * Reads a file containing PEM-encoded X509 certificates and returns a list of parsed
+   * certificates. It splits the file content based on PEM headers and parses each certificate.
+   * Returns an empty list if the trust chain path is empty.
+   *
+   * @param trustChainPath The path to the trust chain file.
+   * @return A list of parsed X509 certificates.
+   * @throws IOException If an error occurs while reading the file.
+   * @throws CertificateException If an error occurs while parsing a certificate.
+   */
+  @VisibleForTesting
+  static List<X509Certificate> readTrustChain(String trustChainPath)
+      throws IOException, CertificateException {
+    List<X509Certificate> certificateTrustChain = new ArrayList<>();
+
+    // If no trust chain path is provided, return an empty list.
+    if (trustChainPath == null || trustChainPath.isEmpty()) {
+      return certificateTrustChain;
+    }
+
+    // initialize certificate factory to retrieve x509 certificates.
+    CertificateFactory cf = CertificateFactory.getInstance("X.509");
+
+    // Read the trust chain file.
+    byte[] trustChainData;
+    try {
+      trustChainData = Files.readAllBytes(Paths.get(trustChainPath));
+    } catch (NoSuchFileException e) {
+      throw new IOException("Trust chain file not found: " + trustChainPath, e);
+    } catch (IOException e) {
+      throw new IOException("Failed to read trust chain file: " + trustChainPath, e);
+    }
+
+    // Split the file content into PEM certificate blocks.
+    String content = new String(trustChainData);
+    Pattern pemCertPattern =
+        Pattern.compile("-----BEGIN CERTIFICATE-----.*?-----END CERTIFICATE-----", Pattern.DOTALL);
```

Comment:
Can the `Pattern` object be a static final constant? This would be compiling the regex every time  `readTrustChain()` is called.

Reply:
Should the inner regex just be `.*`? Just to try an capture where there is the cert details between the `---CERTIFICATE---` blocks? I think it should fail during the parser below if it doesn't match.

Reply:
I think`.*` can lead to problems when we have multiple certificates concatenated in the trust chain file. 

If I'm understanding correctly, Using `.*` (greedy) would match from the first `-----BEGIN CERTIFICATE-----` all the way to the very last `-----END CERTIFICATE-----` in the file, effectively treating all certificates as one large chunk. This would prevent us from processing each certificate individually.

The ? in `.*?` makes the quantifier non-greedy, ensuring it matches the shortest possible string that still allows the `-----END CERTIFICATE-----` to match. This way we can extract each individual certificate block.

Reply:
I see, I think you're right. Thanks for the explanation!